server/jetty/src/main/java/io/deephaven/server/jetty/jsplugin/JsPlugins.java

```diff
@@ -0,0 +1,112 @@
+import io.deephaven.plugin.type.JsPluginRegistration;
+import org.eclipse.jetty.servlet.DefaultServlet;
+import org.eclipse.jetty.servlet.ServletHolder;
+
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.URI;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * A {@link JsPluginRegistration} for Jetty.
+ *
+ * @see #servletHolder(String)
+ */
+public final class JsPlugins implements JsPluginRegistration {
+
+    private static final String PLUGINS = "plugins";
+    private static final String NAME = "name";
+    private static final String VERSION = "version";
+    private static final String MAIN = "main";
+    private static final String MANIFEST_JSON = "manifest.json";
+
+    /**
+     * Creates a new js plugins instance with a temporary zip filesystem.
+     *
+     * @return the js plugins
+     * @throws IOException if an I/O exception occurs
+     */
+    public static JsPlugins create() throws IOException {
+        final Path tempDir = Files.createTempDirectory(JsPlugins.class.getName());
+        tempDir.toFile().deleteOnExit();
+        final Path fsZip = tempDir.resolve("fs.zip");
+        fsZip.toFile().deleteOnExit();
+        // Note, the URI needs explicitly be parseable as a directory URL ending in "!/", a requirement of the jetty
+        // resource creation implementation, see
+        // org.eclipse.jetty.util.resource.Resource.newResource(java.lang.String, boolean)
+        final URI uri = URI.create(String.format("jar:file:%s!/", fsZip));
+        final JsPlugins jsPlugins = new JsPlugins(uri);
+        jsPlugins.init();
+        return jsPlugins;
+    }
+
+    private final URI filesystem;
+    private final List<Map<String, String>> plugins;
+
+    private JsPlugins(URI filesystem) {
+        this.filesystem = Objects.requireNonNull(filesystem);
+        this.plugins = new ArrayList<>();
+    }
+
+    /**
+     * Creates a {@link DefaultServlet} servlet holder.
+     *
+     * @param name the name
+     * @return the servlet holder
+     */
+    public ServletHolder servletHolder(String name) {
+        final ServletHolder jsPlugins = new ServletHolder(name, DefaultServlet.class);
+        jsPlugins.setInitParameter("resourceBase", filesystem.toString());
+        jsPlugins.setInitParameter("pathInfoOnly", "true");
+        jsPlugins.setInitParameter("dirAllowed", "false");
+        // etag support picked up via io.deephaven.server.jetty.JettyBackedGrpcServer
+        // jsPlugins.setInitParameter("etags", "true");
+        jsPlugins.setAsyncSupported(true);
+        return jsPlugins;
+    }
+
+    @Override
+    public synchronized void register(JsPlugin jsPlugin) throws IOException {
+        for (Map<String, String> plugin : plugins) {
```

Comment:
This structure confuses me a bit - why a map here, if there will always and only be three sets of key/value pairs? And then, why require walking the list, if the list is effectively a map from name to object (then serialize map.values() to get the actual underlying values without the extra name wrapping)?

Also, if the structure exists only for json's sake and not for any actual processing, why not use the node types that jackson provides instead of making it adapt from generic collections?

Reply:
it looks like PackageJson is nearly correct for this use case, though perhaps with a name change to be clear that it is about packaging, as opposed to being a subset of the actual package.json specification.

Reply:
I've created a common object called `JsPluginInfo` that can be both ser/deser from JSON.